modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Graph.scala

```diff
@@ -36,29 +36,55 @@ final class Graph private[rdf] (private val underlying: G[Node, LkDiEdge]) {
   def isConnected: Boolean =
     underlying.isConnected
 
+  /**
+    * @param p the triple predicate
+    * @param o the triple object
+    * @return the subjects found from the provided predicate and object
+    */
+  def subjects(p: IriNode, o: Node): Set[IriOrBNode] =
+    underlying.edges.filter(e => p == e.predicate && o == e.obj).map(_.subject).toSet
+
   /**
     * @return the set of nodes in subject position
     */
   def subjects: Set[IriOrBNode] =
-    foldLeft(Set.empty[IriOrBNode]) {
-      case (acc, (s, _, _)) => acc + s
-    }
+    triples.map { case (s, _, _) => s }
+
+  /**
+    * @param s the triple subject
+    * @param o the triple object
+    * @return the predicates found from the provided subject and object
+    */
+  def predicates(s: IriOrBNode, o: Node): Set[IriNode] =
+    underlying.edges.filter(e => s == e.subject && o == e.obj).map(_.predicate).toSet
 
   /**
     * @return the set of predicates
     */
   def predicates: Set[IriNode] =
-    foldLeft(Set.empty[IriNode]) {
-      case (acc, (_, p, _)) => acc + p
-    }
+    triples.map { case (_, p, _) => p }
 
   /**
     * @return the set of nodes in object position
     */
   def objects: Set[Node] =
-    foldLeft(Set.empty[Node]) {
-      case (acc, (_, _, o)) => acc + o
-    }
+    triples.map { case (_, _, o) => o }
+
+  /**
+    * @param s the triple subject
+    * @param p the triple predicate
+    * @return the objects found from the provided subject and predicate
+    */
+  def objects(s: IriOrBNode, p: IriNode): Set[Node] =
+    objects(s2 => s2 == s, p2 => p2 == p)
+
+  /**
+    * @param s the triple subject used to test matches
+    * @param p the triple predicate used to test matches
+    * @return the objects found from the provided subject and predicate
+    */
+  def objects(s: IriOrBNode => Boolean = (n) => true, p: IriNode => Boolean = (n) => true): Set[Node] =
```

Comment:
I would recommend generalising with something like this:
```
  def select(s: IriOrBNode => Boolean, p: IriNode => Boolean, o: Node => Boolean): Set[Triple] = ???
```
and then define all other methods in terms of `select`:
```
  def objects(s: IriOrBNode, p: IriNode): Set[Node] =
    select(_ == s, _ == p).map(_._3)
```
Note: I've recommended `select` as a name because it probably conveys better the action than the standard `filter` name, but we could define an alias if we are to remain compatible with scala collections naming.